### PR TITLE
H/slice layer

### PR DESCRIPTION
**Pull request by hansely**

### CODE REVIEW

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 1]
```diff
@@ -1,345 +1,386 @@
...
+        return -1;
+    }
+    for(int i=0; i<8; i++) {
+        input.push_back((float)fptr[i]);
+    }
+    status = vxUnmapTensorPatch((vx_tensor)parameters[0], map_id);
+
+    // create tmp_input and copy input tensor
+    vx_tensor tmpi_tensor = vxCreateTensor(context, input_num_dims, input_dims, input_type, 0);
+    status = vxMapTensorPatch(tmpi_tensor, input_num_dims, nullptr, nullptr, &map_id, stride, (void **)&fptr, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST, 0);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:122]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170&bid=19436558)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 2]
```diff
@@ -1,345 +1,386 @@
...
+    }
+    for(int i=0; i<icount; i++) {
+        fptr[i] = input[i];
+    }
+    status = vxUnmapTensorPatch(tmpi_tensor, map_id);
+
+    input_tensor[0]  = tmpi_tensor;
+
+    // copy starts index
+    status = vxMapTensorPatch((vx_tensor)parameters[2], 1, nullptr, nullptr, &map_id, stride, (void **)&ptr, VX_READ_ONLY, VX_MEMORY_TYPE_HOST, 0);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:135]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 3]
```diff
@@ -1,345 +1,386 @@
...
+    status = vxMapTensorPatch((vx_tensor)parameters[1], 2, nullptr, nullptr, &map_id, stride, (void **)&fptr, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST, 0);
+    if(status) {
+        std::cerr << "ERROR: vxMapTensorPatch() failed for input3 tensor (" << status << ")" << std::endl;
+        return -1;
+    }
+
+    for(int i=0; i<ocount; i++) {
+        fptr[i] = output[i];
+    }
+    status = vxUnmapTensorPatch((vx_tensor)parameters[1], map_id);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:341]: Codacy found an issue: [Variable 'status' is assigned a value that is never used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 4]
```diff
@@ -1,345 +1,386 @@
...
+    vx_tensor tmpo_tensor = vxCreateTensor(context, output_num_dims, output_dims, input_type, 0);
+    output_tensor[param_dims-1] = tmpo_tensor;
+
+    for (int i=0; i<param_dims; i++) {
+        vxGatherLayer(graph, (vx_tensor)input_tensor[i], (vx_tensor)indices_tensor[i], (vx_tensor)output_tensor[i], axes[i]);
+    }
+    
+    status = vxVerifyGraph(graph);
+    
+    status = vxProcessGraph(graph);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:313]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 5]
```diff
@@ -1,345 +1,386 @@
...
+    
+    // create temp output tensor which will hold the final output result
+    vx_tensor tmpo_tensor = vxCreateTensor(context, output_num_dims, output_dims, input_type, 0);
+    output_tensor[param_dims-1] = tmpo_tensor;
+
+    for (int i=0; i<param_dims; i++) {
+        vxGatherLayer(graph, (vx_tensor)input_tensor[i], (vx_tensor)indices_tensor[i], (vx_tensor)output_tensor[i], axes[i]);
+    }
+    
+    status = vxVerifyGraph(graph);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:311]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 6]
```diff
@@ -1,345 +1,386 @@
...
+        return -1;
+    }
+    
+    for(int i=0; i<ocount; i++) {
+        output.push_back((float)fptr[i]);
+    }
+    status = vxUnmapTensorPatch(tmpo_tensor, map_id);
+ 
+    // write final output
+    status = vxMapTensorPatch((vx_tensor)parameters[1], 2, nullptr, nullptr, &map_id, stride, (void **)&fptr, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST, 0);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:332]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170&bid=19436558)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 7]
```diff
@@ -1,345 +1,393 @@
...
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:339]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 8]
```diff
@@ -1,345 +1,393 @@
...
-        vx_size output_dims[4];
-        ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[i], VX_TENSOR_DIMS, output_dims, sizeof(output_dims)));
-        op_size_per_batch[i-1] = output_dims[2] * output_dims[1] * output_dims[0];
-#if ENABLE_DEBUG_PRINT_DIMS
-        std::cout << "slice output " << i << " " << output_dims[3] << " " << output_dims[2] << " " << output_dims[1] << " " << output_dims[0] << std::endl;
-#endif
-
-        i++;
+    // copy starts index
+    status = vxMapTensorPatch((vx_tensor)parameters[2], 1, nullptr, nullptr, &map_id, stride, (void **)&ptr, VX_READ_ONLY, VX_MEMORY_TYPE_HOST, 0);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:142]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 9]
```diff
@@ -1,345 +1,393 @@
...
+    status = vxMapTensorPatch((vx_tensor)parameters[1], 2, nullptr, nullptr, &map_id, stride, (void **)&fptr, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST, 0);
+    if(status) {
+        std::cerr << "ERROR: vxMapTensorPatch() failed for input3 tensor (" << status << ")" << std::endl;
+        return -1;
+    }
+
+    for(int i=0; i<ocount; i++) {
+        fptr[i] = output[i];
+    }
+    status = vxUnmapTensorPatch((vx_tensor)parameters[1], map_id);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:348]: Codacy found an issue: [Variable 'status' is assigned a value that is never used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 10]
```diff
@@ -1,345 +1,393 @@
...
+    
+    // create temp output tensor which will hold the final output result
+    vx_tensor tmpo_tensor = vxCreateTensor(context, output_num_dims, output_dims, input_type, 0);
+    output_tensor[param_dims-1] = tmpo_tensor;
+
+    for (int i=0; i<param_dims; i++) {
+        vxGatherLayer(graph, (vx_tensor)input_tensor[i], (vx_tensor)indices_tensor[i], (vx_tensor)output_tensor[i], axes[i]);
+    }
+    
+    status = vxVerifyGraph(graph);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:318]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 11]
```diff
@@ -1,345 +1,393 @@
...
+    
+    status = vxProcessGraph(graph);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:320]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: amd_openvx_extensions/amd_nn/src/slice_layer.cpp [hunk 12]
```diff
@@ -1,345 +1,393 @@
...
+        return -1;
+    }
+    for(int i=0; i<8; i++) {
+        input.push_back((float)fptr[i]);
+    }
+    status = vxUnmapTensorPatch((vx_tensor)parameters[0], map_id);
+
+    // create tmp_input and copy input tensor
+    vx_tensor tmpi_tensor = vxCreateTensor(context, input_num_dims, input_dims, input_type, 0);
+    status = vxMapTensorPatch(tmpi_tensor, input_num_dims, nullptr, nullptr, &map_id, stride, (void **)&fptr, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST, 0);
```
Comment (kiritigowda) [inline amd_openvx_extensions/amd_nn/src/slice_layer.cpp:129]: Codacy found an issue: [Variable 'status' is reassigned a value before the old one has been used.](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: model_compiler/python/nnir.py [hunk 13]
```diff
@@ -447,13 +447,24 @@ def updateLocals(self):
...
                     input = self.tensor_dict[node.inputs[0]]
-                    shape = [input.shape[0], input.shape[1] // len(node.outputs), input.shape[2], input.shape[3]]
-                    for name in node.outputs:
-                        local = IrTensor()
-                        local.setName(name)
-                        local.setInfo(input.type, shape)
-                        local.setFormat(input.format)
-                        self.addLocal(local)
+                    starts = np.frombuffer(self.binaries[node.inputs[1]], dtype=np.int32)
+                    ends = np.frombuffer(self.binaries[node.inputs[2]], dtype=np.int32)
```
Comment (kiritigowda) [inline model_compiler/python/nnir.py:451]: Codacy found an issue: [local variable 'ends' is assigned to but never used (F841)](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 3
           

Complexity increasing per file
==============================
- model_compiler/python/nnir.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170&bid=19436558)

File: model_compiler/python/nnir.py [hunk 14]
```diff
@@ -447,13 +447,24 @@ def updateLocals(self):
...
+                    if node.inputs[3]:
+                        axes = np.frombuffer(self.binaries[node.inputs[3]], dtype=np.int32)
+                    else:
+                        axes = [0,1,2,3]
+                    
+                    if node.inputs[4]:
+                        steps = np.frombuffer(self.binaries[node.inputs[4]], dtype=np.int32)
+                    else:
+                        steps = [1,1,1,1]
```
Comment (kiritigowda) [inline model_compiler/python/nnir.py:460]: Codacy found an issue: [local variable 'steps' is assigned to but never used (F841)](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

File: model_compiler/python/nnir.py [hunk 15]
```diff
@@ -447,13 +447,24 @@ def updateLocals(self):
...
                 elif node.type in ['slice']:
                     input = self.tensor_dict[node.inputs[0]]
-                    shape = [input.shape[0], input.shape[1] // len(node.outputs), input.shape[2], input.shape[3]]
-                    for name in node.outputs:
-                        local = IrTensor()
-                        local.setName(name)
-                        local.setInfo(input.type, shape)
-                        local.setFormat(input.format)
-                        self.addLocal(local)
+                    starts = np.frombuffer(self.binaries[node.inputs[1]], dtype=np.int32)
```
Comment (kiritigowda) [inline model_compiler/python/nnir.py:450]: Codacy found an issue: [local variable 'starts' is assigned to but never used (F841)](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5932170)

**Comment by LakshmiKumar23** [comment]
@kiritigowda @hansely Checked output with ONNX examples. Can be merged

**Comment by hansely** [comment]
LGMT warning are resolved for now. It will be enhanced once the whole model is ready.

**Pull request closed without merging**